rust/common-graph/src/error.rs

```diff
@@ -0,0 +1,106 @@
+use crate::PortType;
+use thiserror::Error;
+
+#[cfg(doc)]
+use crate::GraphBuilder;
+
+/// Details of a port in a graph, containing
+/// the node label, port name, and type of port (input/output).
+#[derive(PartialEq, Debug)]
+pub struct PortDetails {
+    /// Label of the node.
+    pub node: String,
+    /// Name of the port.
+    pub port: String,
+    /// Type of port.
+    pub port_type: PortType,
+}
+
+/// Details of a graph connection: the node and port
+/// of both the connection source and target.
+#[derive(PartialEq, Debug)]
+pub struct ConnectionDetails {
+    /// Label of source node.
+    pub source_node: String,
+    /// Name of source port.
+    pub source_port: String,
+    /// Label of target node.
+    pub target_node: String,
+    /// Name of target port.
+    pub target_port: String,
+}
+
+/// Result type for [CommonGraphError].
+pub type Result<T> = ::core::result::Result<T, CommonGraphError>;
+
+/// Errors for graph integrity issues.
+#[derive(PartialEq, Error, Debug)]
+pub enum CommonGraphError {
+    /// There was an invalid cycle in the graph, backtracking
+    /// to a node outside of the root node.
+    #[error("An invalid cycle was detected in the graph.")]
+    InvalidCycleDetected(Box<ConnectionDetails>),
+
+    /// There was either no root nodes, multiple root nodes,
+    /// or a root node not first in the definition.
+    #[error("There must be a single root node in the 0th position.")]
+    InvalidRoot,
+
+    /// There were no nodes found in the graph.
+    #[error("No nodes found in graph.")]
+    EmptyGraph,
+
+    /// There was a reference to a node not found in the graph.
+    #[error("No node with index '{0}' found in graph.")]
+    MissingNode(usize),
+
+    /// There was a reference to a port not found in a node.
+    #[error("No {} port '{}' found in node '{}'.", .0.port_type, .0.port, .0.node)]
+    MissingPort(Box<PortDetails>),
```

Comment:
Helps keep the size of the `Result` small, and `thiserror` supports printing out the inner types anyway